Missing info in activities after running integration for couple of hours

## This is a...
<!-- Check ONLY one of the following options with "x" -->
<pre><code>
[ ] Feature request
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ x] Bug report  <!-- Please search GitHub for a similar issue or PR before submitting -->
[ ] Documentation issue or request
</code></pre>
<!-- If possible, please choose the appropriate labels for your issue. You find a description of all
labels used at https://doc.syndesis.io/#dev-labels -->

## The problem
I have scenario consist of 6 integration with 3messages/s in total. After couple of hours of running (~4 hours 60000 messages)  I'm missing steps infromations in activities:
No steps:
<img width="1228" alt="screen shot 2018-08-07 at 16 39 02" src="https://user-images.githubusercontent.com/6814482/43782945-c7812eb2-9a60-11e8-994e-79722958f0b0.png">
<img width="1211" alt="screen shot 2018-08-07 at 16 42 34" src="https://user-images.githubusercontent.com/6814482/43783000-ecba83d6-9a60-11e8-828f-d64917ffe98d.png">



## Expected behavior
<!-- Describe what the desired behavior would be, enlistin gthe acceptance criteria. -->
always should be all steps visible: 
<img width="1176" alt="screen shot 2018-08-07 at 16 38 53" src="https://user-images.githubusercontent.com/6814482/43783032-011ab9f4-9a61-11e8-9ae0-2b7a77e373fb.png">

## Screenshot
<!--
For features/bugs tackling with UI functionality, drag and drop a screenshot depicting the desired presentation layer or supporting the UX narrative for the new functionality.
-->

## Request and Response Data
<!--
Many issues involve both the UI and it's backend, if possible capture relevant request and response data JSON messages and include it here.
Request and response data can be gathered from your browser's developer tools on the 'Network' tab.
 + As you reproduce the issue, take note of any network requests that are made.
 + Requests that result in an error will be highlighted red.
 + Click on line in the network tab and then the 'Headers' tab to get the request data
 + Click on the 'Preview' or 'Response' tabs to get the response data.
 + Pretty print the json too -> http://jsonprettyprint.com/

BE CAREFUL NOT TO INCLUDE ANY USER TOKENS!!!!
Things like connection objects can contain sensitive data in their configuration, make sure to rip these out
-->

## API Endpoints and Schemas
<!--
For features or bugfixes entailing data exchanges between the UI and the REST API,
enlist the different endpoints available and the payload/response schemas.
-->

## Tasks involved / Steps to Reproduce
<!--
Enlist all the acceptance criteria for new features or the steps required to reproduce the bug/regression reported.
-->
1.
2.
3.
4.

I think the `jstack` output is the most helpful here. What I deduce from it is that the {{eventQueue}} {{LinkedBlockingDeque}} is full at 1000 capacity we allocate for events and that the {{ActivityTrackingController::processEventQueue}} is not running.

{{ActivityTrackingController::processEventQueue}} is started on a separate thread here:
https://github.com/syndesisio/syndesis/blob/dd07eddd58b8520a60edb1de182714bae1eeed3e/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java#L101

And it should be running, until `stopped` flag is set to `true` or an exception is encountered.

From the server-pod logs, it seems that a exception has occurred while processing events:
```
2018-08-07 14:59:36.166 ERROR [-,,,] 1 --- [ntroller [idle]] i.s.s.l.j.c.ActivityTrackingController   : Failure running activity tracking task on thread: Logs Controller [idle]

org.skife.jdbi.v2.exceptions.CallbackFailedException: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.BatchUpdateException: Batch entry 513 INSERT into jsondb (path, value, ovalue) values ('/activity/exchanges/i-LJJ0nl6CuuPiXONnP6Wz/i-LJJZ9kfN2_JW-giX2Ofz/', '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', NULL) ON CONFLICT (path) DO UPDATE SET value = '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', ovalue = NULL was aborted: ERROR: could not extend file "base/16385/16424": No space left on device
  Hint: Check free disk space.  Call getNextException to see other errors in the batch. [statement:"INSERT into jsondb (path, value, ovalue) values (:path, :value, :ovalue) ON CONFLICT (path) DO UPDATE SET value = :value, ovalue = :ovalue", located:"null", rewritten:"null", arguments:{ positional:{}, named:{path:'/activity/pods/i-stats-guess-6-lp9zq/time/',ovalue:'null',value:'`2018-08-07T13:42:26.87125854Z'}, finder:[]}]
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:301) ~[jdbi-2.78.jar!/:2.78]
	at org.skife.jdbi.v2.DBI.inTransaction(DBI.java:346) ~[jdbi-2.78.jar!/:2.78]
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.writeBatch(ActivityTrackingController.java:163) ~[server-logging-jsondb-1.5-SNAPSHOT.jar!/:1.5-SNAPSHOT]
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.processEventQueue(ActivityTrackingController.java:368) ~[server-logging-jsondb-1.5-SNAPSHOT.jar!/:1.5-SNAPSHOT]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_151]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_151]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_151]
Caused by: org.skife.jdbi.v2.exceptions.UnableToExecuteStatementException: java.sql.BatchUpdateException: Batch entry 513 INSERT into jsondb (path, value, ovalue) values ('/activity/exchanges/i-LJJ0nl6CuuPiXONnP6Wz/i-LJJZ9kfN2_JW-giX2Ofz/', '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', NULL) ON CONFLICT (path) DO UPDATE SET value = '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', ovalue = NULL was aborted: ERROR: could not extend file "base/16385/16424": No space left on device
  Hint: Check free disk space.  Call getNextException to see other errors in the batch. [statement:"INSERT into jsondb (path, value, ovalue) values (:path, :value, :ovalue) ON CONFLICT (path) DO UPDATE SET value = :value, ovalue = :ovalue", located:"null", rewritten:"null", arguments:{ positional:{}, named:{path:'/activity/pods/i-stats-guess-6-lp9zq/time/',ovalue:'null',value:'`2018-08-07T13:42:26.87125854Z'}, finder:[]}]
	at org.skife.jdbi.v2.PreparedBatch.internalBatchExecute(PreparedBatch.java:206) ~[jdbi-2.78.jar!/:2.78]
	at org.skife.jdbi.v2.PreparedBatch.execute(PreparedBatch.java:100) ~[jdbi-2.78.jar!/:2.78]
	at io.syndesis.server.logging.jsondb.controller.ActivityTrackingController.lambda$writeBatch$2(ActivityTrackingController.java:191) ~[server-logging-jsondb-1.5-SNAPSHOT.jar!/:1.5-SNAPSHOT]
	at org.skife.jdbi.v2.tweak.transactions.LocalTransactionHandler.inTransaction(LocalTransactionHandler.java:183) ~[jdbi-2.78.jar!/:2.78]
	at org.skife.jdbi.v2.BasicHandle.inTransaction(BasicHandle.java:338) ~[jdbi-2.78.jar!/:2.78]
	at org.skife.jdbi.v2.DBI$5.withHandle(DBI.java:350) ~[jdbi-2.78.jar!/:2.78]
	at org.skife.jdbi.v2.DBI.withHandle(DBI.java:298) ~[jdbi-2.78.jar!/:2.78]
	... 6 common frames omitted
Caused by: java.sql.BatchUpdateException: Batch entry 513 INSERT into jsondb (path, value, ovalue) values ('/activity/exchanges/i-LJJ0nl6CuuPiXONnP6Wz/i-LJJZ9kfN2_JW-giX2Ofz/', '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', NULL) ON CONFLICT (path) DO UPDATE SET value = '`{"id":"i-LJJZ9kfN2_JW-giX2Ofz","logts":"2018-08-07T13:25:50.25218068Z","at":1533648350251,"pod":"i-stats-guess-6-lp9zq","ver":"1","status":"done","failed":false,"steps":[{"id":"-LJJ0i_KEr295bPBIeiG","at":1533648350252,"duration":188769}]}', ovalue = NULL was aborted: ERROR: could not extend file "base/16385/16424": No space left on device
  Hint: Check free disk space.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:145) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.ResultHandlerDelegate.handleError(ResultHandlerDelegate.java:50) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2156) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.v3.QueryExecutorImpl.flushIfDeadlockRisk(QueryExecutorImpl.java:1265) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.v3.QueryExecutorImpl.sendQuery(QueryExecutorImpl.java:1290) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:446) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:793) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1659) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at sun.reflect.GeneratedMethodAccessor190.invoke(Unknown Source) ~[na:na]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_151]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_151]
	at org.apache.tomcat.jdbc.pool.StatementFacade$StatementProxy.invoke(StatementFacade.java:114) ~[tomcat-jdbc-8.5.31.jar!/:na]
	at com.sun.proxy.$Proxy159.executeBatch(Unknown Source) ~[na:na]
	at org.skife.jdbi.v2.PreparedBatch.internalBatchExecute(PreparedBatch.java:196) ~[jdbi-2.78.jar!/:2.78]
	... 12 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: could not extend file "base/16385/16424": No space left on device
  Hint: Check free disk space.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155) ~[postgresql-9.4.1212.jre7.jar!/:9.4.1212.jre7]
	... 23 common frames omitted
```

This suggest that the PostgreSQL database `data` directory, a persistent volume with 1GB capacity has run out of disk space.

--- a/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/support/SupportUtilTest.java
+++ b/app/server/endpoint/src/test/java/io/syndesis/server/endpoint/v1/handler/support/SupportUtilTest.java
@@ -44,7 +44,6 @@ import io.fabric8.openshift.client.dsl.BuildConfigResource;
 import io.fabric8.openshift.client.dsl.DeployableScalableResource;
 import io.fabric8.openshift.client.server.mock.OpenShiftServer;
 import io.syndesis.common.model.EmptyListResult;
-import io.syndesis.common.model.ListResult;
 import io.syndesis.common.model.integration.IntegrationOverview;
 import io.syndesis.server.endpoint.v1.handler.integration.IntegrationHandler;
 import io.syndesis.server.endpoint.v1.handler.integration.support.IntegrationSupportHandler;

--- a/app/server/logging/jsondb/pom.xml
+++ b/app/server/logging/jsondb/pom.xml
@@ -115,6 +115,11 @@
     </dependency>
 
     <dependency>
+      <groupId>commons-io</groupId>
+      <artifactId>commons-io</artifactId>
+    </dependency>
+
+    <dependency>
       <groupId>org.hibernate</groupId>
       <artifactId>hibernate-validator</artifactId>
       <scope>runtime</scope>

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
@@ -49,7 +49,6 @@ import io.fabric8.kubernetes.api.model.PodList;
 import io.fabric8.kubernetes.client.KubernetesClient;
 import io.syndesis.common.util.DurationConverter;
 import io.syndesis.common.util.Json;
-import io.syndesis.common.util.KeyGenerator;
 import io.syndesis.server.jsondb.GetOptions;
 import io.syndesis.server.jsondb.JsonDB;
 import io.syndesis.server.jsondb.impl.JsonRecordSupport;
@@ -80,7 +79,10 @@ public class ActivityTrackingController implements Closeable {
     protected final LinkedBlockingDeque<BatchOperation> eventQueue = new LinkedBlockingDeque<>(1000);
     protected final AtomicBoolean stopped = new AtomicBoolean();
 
-    private Duration retention = Duration.ofDays(1);
+    /** The number of retained activity log items */
+    private int retention = 50;
+    /** Logs older than this will not be processed */
+    private Duration retentionTime = Duration.ofDays(1);
     private Duration cleanUpInterval = Duration.ofMinutes(15);
     private Duration startupDelay = Duration.ofSeconds(15);
     private SqlJsonDB.DatabaseKind databaseKind;
@@ -126,16 +128,14 @@ public class ActivityTrackingController implements Closeable {
         Thread.currentThread().setName("Logs Controller Scheduler [running]: cleanupLogs");
 
         try {
-            LOG.info("Purging old controller");
-            long until = System.currentTimeMillis() - retention.toMillis();
-            String untilKey = KeyGenerator.recreateKey(until, 0, 0);
+            LOG.info("Purging old activity logs");
 
             @SuppressWarnings("unchecked")
             Map<String, Object> hashMap = dbGet(HashMap.class, "/activity/integrations"); //NOPMD
             if( hashMap!=null ) {
                 for (String integrationId : hashMap.keySet()) {
-                    String integrationPath = "/activity/exchanges/" + integrationId + "/";
-                    int count = deleteFieldsLT(integrationPath, untilKey);
+                    String integrationPath = "/activity/exchanges/" + integrationId + "/%";
+                    int count = deleteKeepingRetention(integrationPath);
                     LOG.info("deleted {} transactions for integration: {}", count, integrationId);
                 }
             }
@@ -152,19 +152,28 @@ public class ActivityTrackingController implements Closeable {
      * @param path
      * @param field
      */
-    private int deleteFieldsLT(String path, String field) {
+    int deleteKeepingRetention(String path) {
         return dbi.inTransaction((conn, status) -> {
-            StringBuilder sql = new StringBuilder("DELETE from jsondb where path LIKE ? and path < ?");
-            return conn.update(sql.toString(), path+"%", path + field);
+            final String sql = "DELETE FROM jsondb "
+                + "WHERE path LIKE ? AND path NOT IN ("
+                +     "SELECT path FROM jsondb WHERE path LIKE ? ORDER BY path FETCH FIRST (?) ROWS ONLY"
+                + ")";
+
+                return conn.update(sql, path, path, retention);
         });
     }
 
     private void writeBatch(Map<String, Object> batch) {
         dbi.inTransaction((conn, status) -> {
-            String sql = "INSERT into jsondb (path, value, ovalue) values (:path, :value, :ovalue)";
+            final String sql;
             if( databaseKind == SqlJsonDB.DatabaseKind.PostgreSQL ) {
                 // Lets update if the record exists.
-                sql += " ON CONFLICT (path) DO UPDATE SET value = :value, ovalue = :ovalue"; // NOPMD
+                sql =  "INSERT into jsondb (path, value, ovalue) values (:path, :value, :ovalue) "
+                        + "ON CONFLICT (path) DO UPDATE SET value = :value, ovalue = :ovalue";
+            } else if (databaseKind == SqlJsonDB.DatabaseKind.H2) {
+                sql =  "MERGE INTO jsondb (path, value, ovalue) VALUES (:path, :value, :ovalue)";
+            } else {
+                sql = "INSERT into jsondb (path, value, ovalue) values (:path, :value, :ovalue)";
             }
 
             PreparedBatch insert = conn.prepareBatch(sql);
@@ -368,7 +377,7 @@ public class ActivityTrackingController implements Closeable {
                     try {
                         writeBatch(batch);
                     } catch (@SuppressWarnings("PMD.AvoidCatchingGenericException") RuntimeException e) {
-                        LOG.warn("Unable to write batch of events: ", e.getMessage());
+                        LOG.warn("Unable to write batch of events: {}", e.getMessage());
                         LOG.debug("Unable to write batch of events: ", e);
                     }
 
@@ -388,12 +397,22 @@ public class ActivityTrackingController implements Closeable {
 
     @SuppressWarnings("FutureReturnValueIgnored")
     protected void schedule(Runnable command, long delay, TimeUnit unit) {
+        if (stopped.get()) {
+            LOG.warn("Not scheduling command: {}, the activity tracking is stopping", command);
+            return;
+        }
+
         scheduler.schedule(()->{ executor.execute(command); }, delay, unit);
     }
 
-    @Value("${controllers.dblogging.retention:1 day}")
-    public void setRetention(String retention) {
-        this.retention = new DurationConverter().convert(retention);
+    @Value("${controllers.dblogging.retention:50}")
+    public void setRetention(int retention) {
+        this.retention = retention;
+    }
+
+    @Value("${controllers.dblogging.retentionTime:1 day}")
+    public void setRetentionTime(final String retentionTime) {
+        this.retentionTime = new DurationConverter().convert(retentionTime);
     }
 
     @Value("${controllers.dblogging.cleanUpPeriod:15 minutes}")
@@ -406,11 +425,20 @@ public class ActivityTrackingController implements Closeable {
         this.startupDelay = new DurationConverter().convert(startupDelay);
     }
 
-    public Duration getRetention() {
+    public int getRetention() {
         return retention;
     }
 
+    public Duration getRetentionTime() {
+        return retentionTime;
+    }
+
     void execute(String podName, Runnable task) {
+        if (stopped.get()) {
+            LOG.warn("Not executing task: {}, for pod {}, the activity tracking is stopping", task, podName);
+            return;
+        }
+
         executor.execute(() -> {
             Thread.currentThread().setName("Logs Controller [running], pod: " + podName);
             try {

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
@@ -23,6 +23,7 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.Map;
 import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.concurrent.LinkedBlockingDeque;
@@ -69,7 +70,7 @@ public class ActivityTrackingController implements Closeable {
 
     private final DBI dbi;
     private final KubernetesClient client;
-    private final Map<String, PodLogMonitor> podHandlers = new HashMap<>();
+    private final Map<String, PodLogMonitor> podHandlers = new ConcurrentHashMap<>();
     private final JsonDB jsondb;
     private ScheduledExecutorService scheduler;
     private ExecutorService executor;
@@ -247,22 +248,13 @@ public class ActivityTrackingController implements Closeable {
                 }
 
                 String name = pod.getMetadata().getName();
-                PodLogMonitor handler = podHandlers.get(name);
+                PodLogMonitor handler = podHandlers.computeIfAbsent(name, n -> createLogMonitor(pod));
 
-                if (handler == null) {
-                    // create a new handler.
-                    try {
-                        handler = createLogMonitor(pod);
-                        handler.start();
-
-                        LOG.info("Created handler for pod: {}", handler.podName);
-                        podHandlers.put(name, handler);
-                    } catch (IOException e) {
-                        LOG.error("Unexpected Error", e);
-                    }
-                } else {
-                    // mark existing handlers as being used.
+                try {
                     handler.markInOpenshift.set(true);
+                    handler.start();
+                } catch (IOException e) {
+                    LOG.error("Unexpected Error", e);
                 }
             }
 
@@ -294,8 +286,12 @@ public class ActivityTrackingController implements Closeable {
 
     }
 
-    protected PodLogMonitor createLogMonitor(Pod pod) throws IOException {
-        return new PodLogMonitor(this, pod);
+    protected PodLogMonitor createLogMonitor(Pod pod) {
+        final PodLogMonitor monitor = new PodLogMonitor(this, pod);
+
+        LOG.info("Created log monitor for pod: {}", monitor.podName);
+
+        return monitor;
     }
 
     protected PodList listPods() {
@@ -393,16 +389,6 @@ public class ActivityTrackingController implements Closeable {
             Thread.currentThread().setName(IDLE_THREAD_NAME);
         }
         LOG.info("Batch ingestion work thread done.");
-    }
-
-    @SuppressWarnings("FutureReturnValueIgnored")
-    protected void schedule(Runnable command, long delay, TimeUnit unit) {
-        if (stopped.get()) {
-            LOG.warn("Not scheduling command: {}, the activity tracking is stopping", command);
-            return;
-        }
-
-        scheduler.schedule(()->{ executor.execute(command); }, delay, unit);
     }
 
     @Value("${controllers.dblogging.retention:50}")

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingController.java
@@ -365,7 +365,12 @@ public class ActivityTrackingController implements Closeable {
                     }
 
                     // Write the batch..
-                    writeBatch(batch);
+                    try {
+                        writeBatch(batch);
+                    } catch (@SuppressWarnings("PMD.AvoidCatchingGenericException") RuntimeException e) {
+                        LOG.warn("Unable to write batch of events: ", e.getMessage());
+                        LOG.debug("Unable to write batch of events: ", e);
+                    }
 
                     LOG.debug("Batch ingested {} log events", eventCounter);
 

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupport.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupport.java
@@ -69,7 +69,7 @@ public class KubernetesSupport {
                 .append(pod.getResourceUrl().toString())
                 .append("/log?pretty=false&follow=true&timestamps=true");
             if (sinceTime != null) {
-                url.append("&sinceTime=");
+                url.append("&sinceTime=").append(sinceTime);
             }
             String podLogUrl = url.toString();
 

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupport.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupport.java
@@ -91,9 +91,7 @@ public class KubernetesSupport {
                         Thread.currentThread().setName("Logs Controller [running], streaming: " + podLogUrl);
                         try {
                             if (response.code() == 200) {
-                                try (InputStream responseStream = response.body().byteStream()) {
-                                    handler.accept(responseStream);
-                                }
+                                handler.accept(response.body().byteStream());
                             } else {
                                 LOG.info("Failure occurred while processing controller for pod: {}, http status: {}, details: {}", podName, response.code(), response.body().string());
                                 handler.accept(null);

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
@@ -65,7 +65,7 @@ class PodLogMonitor implements Consumer<InputStream> {
     protected PodLogState state;
     protected HashMap<String, InflightData> inflightActivities = new HashMap<>();
 
-    public PodLogMonitor(ActivityTrackingController logsController, Pod pod) throws IOException {
+    PodLogMonitor(ActivityTrackingController logsController, Pod pod) throws IOException {
         this.logsController = logsController;
         this.podName = pod.getMetadata().getName();
         if (this.podName == null) {
@@ -217,7 +217,7 @@ class PodLogMonitor implements Consumer<InputStream> {
             // are the required fields set?
             String exchange = validate((String) json.remove("exchange"));
             long keyTimeMillis = KeyGenerator.getKeyTimeMillis(exchange);
-            long until = System.currentTimeMillis() - logsController.getRetention().toMillis();
+            long until = now() - logsController.getRetentionTime().toMillis();
             if( keyTimeMillis < until ) {
                 // This log entry is too old.. don't process it..
                 return;
@@ -312,6 +312,10 @@ class PodLogMonitor implements Consumer<InputStream> {
             rethrow.initCause(e);
             throw rethrow;
         }
+    }
+
+    long now() {
+        return System.currentTimeMillis();
     }
 
     private static JsonNode toJsonNode(Map<String, Object> json) throws IOException {

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
@@ -29,7 +29,6 @@ import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
-import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
 import java.util.regex.Matcher;
@@ -57,6 +56,7 @@ class PodLogMonitor implements Consumer<InputStream> {
     // matches log lines like: 2018-06-06T21:54:36.30603486Z {"exchange":"i-LEM51uGKc6IuIjvR95Vz","status":"begin"}
     private static final Pattern LOG_LINE_REGEX = Pattern.compile("^(\\d\\d\\d\\d\\-\\d\\d\\-\\d\\dT\\d\\d:\\d\\d:\\d\\d\\.\\d+Z) (\\{.*\\})\\s*");
 
+    private AtomicBoolean running = new AtomicBoolean(false);
     private final ActivityTrackingController logsController;
     protected final AtomicBoolean markInOpenshift = new AtomicBoolean(true);
     protected final AtomicBoolean keepTrying = new AtomicBoolean(true);
@@ -66,26 +66,29 @@ class PodLogMonitor implements Consumer<InputStream> {
     protected PodLogState state;
     protected HashMap<String, InflightData> inflightActivities = new HashMap<>();
 
-    PodLogMonitor(ActivityTrackingController logsController, Pod pod) throws IOException {
+    PodLogMonitor(ActivityTrackingController logsController, Pod pod) {
         this.logsController = logsController;
         this.podName = pod.getMetadata().getName();
         if (this.podName == null) {
-            throw new IOException("Could not determine the pod name");
+            throw new IllegalStateException("Could not determine the pod name");
         }
 
         Map<String, String> labels = pod.getMetadata().getLabels();
         this.integrationId = labels.get(OpenShiftService.INTEGRATION_ID_LABEL);
         if (this.integrationId == null) {
-            throw new IOException("Could not determine the integration id that is being run on the pod: " + this.podName);
+            throw new IllegalStateException("Could not determine the integration id that is being run on the pod: " + this.podName);
         }
 
         this.deploymentVersion = labels.get(OpenShiftService.DEPLOYMENT_VERSION_LABEL);
         if (this.deploymentVersion == null) {
-            throw new IOException("Could not determine the deployment version that is being run on the pod: " + this.podName);
+            throw new IllegalStateException("Could not determine the deployment version that is being run on the pod: " + this.podName);
         }
     }
 
     public void start() throws IOException {
+        if (running.get()) {
+            return;
+        }
 
         // We are just getting started, means we are in the openshift pod
         // list and we need to track pod log state.
@@ -105,31 +108,32 @@ class PodLogMonitor implements Consumer<InputStream> {
             return;
         }
 
-        LOG.info("Getting controller for pod: {}", podName);
-        try {
-            logsController.watchLog(this.podName, this, this.state.time);
-        } catch (IOException e) {
-            LOG.info("Failure occurred while processing controller for pod: {}", podName, e);
-            logsController.schedule(this::run, 5, TimeUnit.SECONDS);
+        if (running.compareAndSet(false, true)) {
+            LOG.info("Getting controller for pod: {}", podName);
+            try {
+                logsController.watchLog(this.podName, this, this.state.time);
+            } catch (IOException e) {
+                LOG.info("Failure occurred while processing controller for pod: {}", podName, e);
+            }
         }
     }
 
     @Override
     public void accept(InputStream is) {
-        if (is != null) {
-            try(InputStream stream = is) {
-                processLogStream(stream);
-            } catch (SocketTimeoutException | EOFException e) {
-                LOG.info("Streaming ended for pod {} due to: {}", podName, message(e));
-                LOG.debug("Streaming ended for pod {}", podName, e);
-            } catch (InterruptedException | IOException e) {
-                LOG.info("Failure occurred while processing controller for pod: {}", podName, e);
-            } finally {
-                logsController.schedule(this::run, 5, TimeUnit.SECONDS);
-                IOUtils.closeQuietly(is);
+        try {
+            if (is == null) {
+                return;
             }
-        } else {
-            logsController.schedule(this::run, 5, TimeUnit.SECONDS);
+
+            processLogStream(is);
+        } catch (SocketTimeoutException | EOFException e) {
+            LOG.info("Streaming ended for pod {} due to: {}", podName, message(e));
+            LOG.debug("Streaming ended for pod {}", podName, e);
+        } catch (InterruptedException | IOException e) {
+            LOG.info("Failure occurred while processing controller for pod: {}", podName, e);
+        } finally {
+            running.set(false);
+            IOUtils.closeQuietly(is);
         }
     }
 
@@ -159,7 +163,6 @@ class PodLogMonitor implements Consumer<InputStream> {
             if (logsController.isPodRunning(podName)) {
                 // odd, why did our stream end??  try to resume processing..
                 LOG.info("End of Log stream for running pod: {}", podName);
-                logsController.schedule(this::run, 5, TimeUnit.SECONDS);
             } else {
                 // Seems like the normal case where stream ends because pod is stopped.
                 LOG.info("End of Log stream for terminated pod: {}", podName);

--- a/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
+++ b/app/server/logging/jsondb/src/main/java/io/syndesis/server/logging/jsondb/controller/PodLogMonitor.java
@@ -35,6 +35,7 @@ import java.util.function.Consumer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
+import org.apache.commons.io.IOUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -125,6 +126,7 @@ class PodLogMonitor implements Consumer<InputStream> {
                 LOG.info("Failure occurred while processing controller for pod: {}", podName, e);
             } finally {
                 logsController.schedule(this::run, 5, TimeUnit.SECONDS);
+                IOUtils.closeQuietly(is);
             }
         } else {
             logsController.schedule(this::run, 5, TimeUnit.SECONDS);

--- a/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingControllerTest.java
+++ b/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingControllerTest.java
@@ -20,20 +20,31 @@ import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
 import java.util.function.Consumer;
+import java.util.function.Supplier;
 
 import static java.util.Objects.requireNonNull;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.fabric8.kubernetes.api.model.Pod;
 import io.fabric8.kubernetes.api.model.PodList;
 import io.fabric8.kubernetes.api.model.PodListBuilder;
 import io.fabric8.kubernetes.client.ConfigBuilder;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.syndesis.common.util.Json;
+import io.syndesis.common.util.KeyGenerator;
 import io.syndesis.server.jsondb.GetOptions;
+import io.syndesis.server.jsondb.JsonDB;
 import io.syndesis.server.jsondb.impl.SqlJsonDB;
 import io.syndesis.server.openshift.OpenShiftService;
 
 import org.h2.jdbcx.JdbcDataSource;
+import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 import org.skife.jdbi.v2.DBI;
@@ -48,8 +59,60 @@ import static org.mockito.Mockito.when;
  */
 public class ActivityTrackingControllerTest {
 
+    private static final class StubbedController extends ActivityTrackingController {
+        private final InputStream podLogs;
+
+        private long time;
+
+        private StubbedController(JsonDB jsondb, DBI dbi, KubernetesClient client, InputStream podLogs, long time) {
+            super(jsondb, dbi, client);
+            this.podLogs = podLogs;
+            this.time = time;
+        }
+
+        @Override
+        protected PodList listPods() {
+            return new PodListBuilder()
+                .addNewItem()
+                .withNewMetadata()
+                .withName("test-pod-x23x")
+                .addToLabels(OpenShiftService.COMPONENT_LABEL, "integration")
+                .addToLabels(OpenShiftService.DEPLOYMENT_VERSION_LABEL, "3")
+                .addToLabels(OpenShiftService.INTEGRATION_ID_LABEL, "my-integration")
+                .endMetadata()
+                .withNewStatus()
+                .withPhase("Running")
+                .endStatus()
+                .endItem().build();
+        }
+
+        @Override
+        protected boolean isPodRunning(String name) {
+            return true;
+        }
+
+        @Override
+        protected void watchLog(String podName, Consumer<InputStream> handler, String sinceTime) throws IOException {
+            execute("test", ()->{
+                handler.accept(podLogs);
+            });
+        }
+
+        @Override
+        protected PodLogMonitor createLogMonitor(Pod pod) throws IOException {
+            return new PodLogMonitor(this, pod) {
+                @Override
+                long now() {
+                    return time;
+                }
+            };
+        }
+    }
+
     private SqlJsonDB jsondb;
     private DBI dbi;
+
+    private KubernetesClient client;
 
     @Before
     public void before() {
@@ -58,50 +121,26 @@ public class ActivityTrackingControllerTest {
         this.dbi = new DBI(ds);
         this.jsondb = new SqlJsonDB(dbi, null);
         this.jsondb.createTables();
+
+        client = mock(KubernetesClient.class);
+        when(client.getConfiguration()).thenReturn(new ConfigBuilder().withMasterUrl("http://master").build());
+    }
+
+    @After
+    public void destroyEverything() {
+        jsondb.dropTables();
     }
 
     @Test
     public void testLogsController() throws IOException {
-
         final String expectedDBState = resource("logs-controller-db.json").trim();
         final String podLogs = resource("test-pod-x23x.txt");
+        final InputStream podLogsStream = new ByteArrayInputStream(podLogs.getBytes(StandardCharsets.UTF_8));
 
-        final KubernetesClient client = mock(KubernetesClient.class);
-        when(client.getConfiguration()).thenReturn(new ConfigBuilder().withMasterUrl("http://master").build());
-        try (ActivityTrackingController controller = new ActivityTrackingController(jsondb, dbi, client) {
-
-            @Override
-            protected PodList listPods() {
-                return new PodListBuilder()
-                    .addNewItem()
-                    .withNewMetadata()
-                    .withName("test-pod-x23x")
-                    .addToLabels(OpenShiftService.COMPONENT_LABEL, "integration")
-                    .addToLabels(OpenShiftService.DEPLOYMENT_VERSION_LABEL, "3")
-                    .addToLabels(OpenShiftService.INTEGRATION_ID_LABEL, "my-integration")
-                    .endMetadata()
-                    .withNewStatus()
-                    .withPhase("Running")
-                    .endStatus()
-                    .endItem().build();
-            }
-
-            @Override
-            protected boolean isPodRunning(String name) {
-                return true;
-            }
-
-            @Override
-            protected void watchLog(String podName, Consumer<InputStream> handler, String sinceTime) throws IOException {
-                execute("test", ()->{
-                    handler.accept(new ByteArrayInputStream(podLogs.getBytes(StandardCharsets.UTF_8)));
-                });
-            }
-
-        }) {
+        try (ActivityTrackingController controller = new StubbedController(jsondb, dbi, client, podLogsStream, ZonedDateTime.parse("2018-01-12T21:22:02.068338027Z").toInstant().toEpochMilli())) {
 
             controller.setStartupDelay("0 seconds");
-            controller.setRetention("1000000000 days");
+            controller.setRetention(Integer.MAX_VALUE);
             controller.open();
 
             // Eventually all the log data should make it into the jsondb
@@ -114,6 +153,76 @@ public class ActivityTrackingControllerTest {
                 });
         }
 
+    }
+
+    @Test
+    public void shouldRetainLastRetainActivityLogs() throws IOException {
+        try (InputStream podLogsStream = logStream(1100);
+            ActivityTrackingController controller = new StubbedController(jsondb, dbi, client, podLogsStream, System.currentTimeMillis())) {
+
+            controller.setStartupDelay("0 seconds");
+            controller.setCleanUpInterval("15 minutes");
+            controller.open();
+
+            // Eventually all the log data should make it into the jsondb
+            given().await()
+                .atMost(20, SECONDS)
+                .pollInterval(1, SECONDS)
+                .untilAsserted(() -> {
+                    final String json = jsondb.getAsString("/", new GetOptions().prettyPrint(true));
+                    assertThat(json).isNotNull();
+                    final JsonNode tree = Json.reader().readTree(json);
+                    assertThat(tree.get("activity").get("exchanges").get("my-integration").size()).isGreaterThan(1000);
+                });
+
+            controller.cleanupLogs();
+            final String json = jsondb.getAsString("/", new GetOptions().prettyPrint(true));
+            assertThat(json).isNotNull();
+            final JsonNode tree = Json.reader().readTree(json);
+            assertThat(tree.get("activity").get("exchanges").get("my-integration").size()).isLessThanOrEqualTo(controller.getRetention());
+        }
+    }
+
+    private static String timestamp() {
+        return DateTimeFormatter.ISO_OFFSET_DATE_TIME.format(ZonedDateTime.now(ZoneOffset.UTC));
+    }
+
+    @SuppressWarnings("InputStreamSlowMultibyteRead")
+    private static InputStream logStream(final int totalExchanges) {
+        final InputStream podLogsStream = new InputStream() {
+            int exchanges;
+
+            int idx;
+
+            byte[] buffy;
+
+            String exchange;
+
+            final Supplier<byte[]> lineSupplier = () ->
+            (timestamp() + " {\"exchange\":\"" + exchange + "\",\"status\":\"begin\"}\n"
+             + timestamp() + " {\"exchange\":\"" + exchange + "\",\"step\":\"s2\",\"id\":\"i-L38cZ5Jd1L876xV4vEHz\",\"message\":\"Hello World\"}\n"
+             + timestamp() + " {\"exchange\":\"" + exchange + "\",\"step\":\"s2\",\"id\":\"i-L38cZ5Jd1L876xV4vEGz\",\"duration\":582977}\n"
+             + timestamp() + " {\"exchange\":\"" + exchange + "\",\"step\":\"s3\",\"id\":\"-L38cZ5Nd1L876xV4vEI\",\"duration\":18087}\n"
+             + timestamp() + " {\"exchange\":\"" + exchange + "\",\"step\":\"s4\",\"id\":\"i-L38cZ5Od1L876xV4vEJz\",\"duration\":494949}\n"
+             + timestamp() + " {\"exchange\":\"" + exchange + "\",\"status\":\"done\",\"failed\":false}\n"
+            ).getBytes(StandardCharsets.UTF_8);
+
+            @Override
+            public int read() throws IOException {
+                if (buffy == null || buffy.length == idx) {
+                    if (exchanges++ > totalExchanges) {
+                        return -1;
+                    }
+
+                    exchange = KeyGenerator.createKey();
+                    buffy = lineSupplier.get();
+                    idx = 0;
+                }
+
+                return buffy[idx++];
+            }
+        };
+        return podLogsStream;
     }
 
     private static String resource(String file) throws IOException {

--- a/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingControllerTest.java
+++ b/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/ActivityTrackingControllerTest.java
@@ -99,7 +99,7 @@ public class ActivityTrackingControllerTest {
         }
 
         @Override
-        protected PodLogMonitor createLogMonitor(Pod pod) throws IOException {
+        protected PodLogMonitor createLogMonitor(Pod pod) {
             return new PodLogMonitor(this, pod) {
                 @Override
                 long now() {

--- a/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupportIT.java
+++ b/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/KubernetesSupportIT.java
@@ -127,7 +127,7 @@ public class KubernetesSupportIT {
                 }
 
                 @Override
-                protected PodLogMonitor createLogMonitor(Pod pod) throws IOException {
+                protected PodLogMonitor createLogMonitor(Pod pod) {
                     return new TestPodLogMonitor(this, pod, counter, phaser);
                 }
             };

--- a/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/TestPodLogMonitor.java
+++ b/app/server/logging/jsondb/src/test/java/io/syndesis/server/logging/jsondb/controller/TestPodLogMonitor.java
@@ -25,7 +25,7 @@ public class TestPodLogMonitor extends PodLogMonitor {
     private final AtomicInteger counter;
     private final Phaser phaser;
 
-    public TestPodLogMonitor(final ActivityTrackingController logsController, final Pod pod, final AtomicInteger counter, final Phaser phaser) throws IOException {
+    public TestPodLogMonitor(final ActivityTrackingController logsController, final Pod pod, final AtomicInteger counter, final Phaser phaser) {
         super(logsController, pod);
         this.counter = counter;
         this.phaser = phaser;

